File: sdk/python/console.py
"""Clutch Python SDK interactive console.

A REPL for driving the SDK by hand: authenticate, inspect tenants and locks, connect the lock
client, and acquire, release, and heartbeat locks while watching server pushes.

Usage: python console.py [endpoint]
"""

import sys

from clutch_admin_sdk import ClutchAdminClient, ClutchError
from clutch_lock_sdk import ClutchLockClient, ClutchLockDeniedError


def print_help():
    print("Commands:")
    print("  login key <accessKey> [secretKey]     Authenticate with an application key")
    print("  login user <tenantId> <email> <pw>    Authenticate with user credentials")
    print("  whoami                                Show the current principal")
    print("  serverinfo                            Show server information")
    print("  tenants                               List tenants")
    print("  locks                                 List active locks in the current tenant")
    print("  connect <accessKey> [secretKey]       Open the WebSocket lock connection")
    print("  acquire <key> <Read|Write|Delete> [wait <ms>]   Acquire a lock")
    print("  held                                  List locks held on this connection")
    print("  release <n>                           Release a held lock by its list number")
    print("  heartbeat                             Send a heartbeat for all held locks")
    print("  logout                                Revoke the current session")
    print("  quit                                  Exit")


def main():
    endpoint = sys.argv[1] if len(sys.argv) > 1 else "http://127.0.0.1:8090"
    admin = ClutchAdminClient(endpoint)
    state = {"tenant_id": None, "locks": None, "counter": 0}
    held = {}

    print("")
    print('Clutch interactive console. Type "help" for commands, "quit" to exit.')
    print(f"Endpoint: {endpoint}")
    print("")

    while True:
        try:
            line = input("clutch> ").strip()
        except EOFError:
            break
        if not line:
            continue
        parts = line.split()
        command = parts[0].lower()

        try:
            if command == "help":
                print_help()
            elif command == "login":
                if len(parts) >= 3 and parts[1].lower() == "key":
                    token = admin.authenticate_with_key(parts[2], parts[3] if len(parts) >= 4 else parts[2])
                    state["tenant_id"] = token.get("tenantId")
                    print(f"Authenticated. Tenant: {token.get('tenantId')}")
                elif len(parts) >= 5 and parts[1].lower() == "user":
                    token = admin.authenticate_with_password(parts[2], parts[3], parts[4])
                    state["tenant_id"] = token.get("tenantId")
                    print(f"Authenticated. Tenant: {token.get('tenantId')}")
                else:
                    print("Usage: login key <accessKey> [secretKey]  |  login user <tenantId> <email> <password>")
            elif command == "whoami":
                d = admin.get_token_details()
                print(f"Principal : {d.get('principalName')}")
                print(f"Type      : {d.get('principalType')}")
                print(f"Tenant    : {d.get('tenantId')}")
                print(f"IsAdmin   : {d.get('isAdmin')}")
                print(f"TenantAdmin: {d.get('isTenantAdmin')}")
            elif command == "serverinfo":
                info = admin.get_server_info()
                print(f"Product : {info.get('product')} {info.get('version')}")
                print(f"Node    : {info.get('node')}")
                print(f"Database: {info.get('database')}")
                print(f"WS conns: {info.get('webSocketConnections')}")
            elif command == "tenants":
                tenants = admin.list_tenants()
                print(f"{len(tenants)} tenant(s):")
                for t in tenants:
                    print(f"  {t.get('id')}  {t.get('name')}  (active={t.get('active')}, protected={t.get('isProtected')})")
            elif command == "locks":
                if not state["tenant_id"]:
                    print('Not authenticated. Use "login" first.')
                else:
                    holders = admin.list_locks(state["tenant_id"])
                    print(f"{len(holders)} active holder(s):")
                    for h in holders:
                        print(f"  {h.get('lockKey')}  {h.get('mode')}  fencing={h.get('fencingToken')}  holder={h.get('id')}")
            elif command == "connect":
                if len(parts) < 2:
                    print("Usage: connect <accessKey> [secretKey]")
                else:
                    if state["locks"] is not None:
                        state["locks"].close()
                    client = ClutchLockClient(endpoint, parts[1], parts[2] if len(parts) >= 3 else None)
                    client.on_heartbeat = lambda renewed: print(f"\n[push] heartbeat renewed {len(renewed)} lease(s)")
                    client.on_error = lambda message: print(f"\n[push] error: {message}")
                    client.on_close = lambda reason: print(f"\n[push] connection closed: {reason}")
                    welcome = client.connect()
                    state["locks"] = client
                    print(f"Connected. Session {welcome.get('sessionId')}, heartbeat every {welcome.get('heartbeatIntervalMs')} ms.")
            elif command == "acquire":
                if state["locks"] is None:
                    print('Not connected. Use "connect" first.')
                elif len(parts) < 3:
                    print("Usage: acquire <key> <Read|Write|Delete> [wait <ms>]")
                else:
                    kwargs = {}
                    if len(parts) >= 5 and parts[3].lower() == "wait":
                        kwargs = {"behavior": "Wait", "timeout_ms": int(parts[4])}
                    acquired = state["locks"].acquire(parts[1], parts[2], **kwargs)
                    state["counter"] += 1
                    n = state["counter"]
                    held[n] = acquired
                    print(f"Acquired [{n}] key={acquired.get('key')} mode={acquired.get('mode')} "
                          f"fencing={acquired.get('fencingToken')} lease={acquired.get('leaseExpiresUtc')}")
            elif command == "held":
                if not held:
                    print("No locks held on this connection.")
                else:
                    for n, h in held.items():
                        print(f"  [{n}] key={h.get('key')} mode={h.get('mode')} fencing={h.get('fencingToken')} holder={h.get('holderId')}")
            elif command == "release":
                if state["locks"] is None:
                    print('Not connected. Use "connect" first.')
                elif len(parts) < 2 or not parts[1].isdigit() or int(parts[1]) not in held:
                    print('Usage: release <n>  (see "held" for numbers)')
                else:
                    n = int(parts[1])
                    released = state["locks"].release(held[n]["holderId"])
                    del held[n]
                    print(f"Released [{n}]: {released}")
            elif command == "heartbeat":
                if state["locks"] is None:
                    print('Not connected. Use "connect" first.')
                else:
                    state["locks"].heartbeat()
                    print("Heartbeat sent for all held locks.")
            elif command == "logout":
                admin.logout()
                state["tenant_id"] = None
                print("Logged out.")
            elif command in ("quit", "exit"):
                break
            else:
                print(f"Unknown command '{command}'. Type 'help'.")
        except ClutchLockDeniedError as ex:
            print(f"Lock denied ({ex.result}): {ex}")
        except ClutchError as ex:
            print(f"Error: {ex}")
        except Exception as ex:  # noqa: BLE001
            print(f"Unexpected: {type(ex).__name__}: {ex}")

    if state["locks"] is not None:
        state["locks"].close()
    admin.close()
    print("Goodbye.")


if __name__ == "__main__":
    main()

File: sdk/js/console.js
'use strict';

/**
 * Clutch JavaScript SDK interactive console.
 *
 * A readline REPL for driving the SDK by hand: authenticate, inspect tenants and locks, connect the
 * lock client, and acquire, release, and heartbeat locks while watching server pushes.
 *
 * Usage: node console.js [endpoint]   (or: npm run console)
 */

const readline = require('readline');
const { ClutchAdminClient } = require('./clutch-admin-sdk');
const { ClutchLockClient, ClutchLockDeniedError } = require('./clutch-lock-sdk');

const endpoint = process.argv[2] || 'http://127.0.0.1:8090';
const admin = new ClutchAdminClient(endpoint);
let locks = null;
let tenantId = null;
const held = new Map();
let holderCounter = 0;

const rl = readline.createInterface({ input: process.stdin, output: process.stdout, prompt: 'clutch> ' });

function printHelp() {
    console.log('Commands:');
    console.log('  login key <accessKey>                 Authenticate with an application key');
    console.log('  login user <tenantId> <email> <pw>    Authenticate with user credentials');
    console.log('  whoami                                Show the current principal');
    console.log('  serverinfo                            Show server information');
    console.log('  tenants                               List tenants');
    console.log('  locks                                 List active locks in the current tenant');
    console.log('  connect <accessKey>                   Open the WebSocket lock connection');
    console.log('  acquire <key> <Read|Write|Delete> [wait <ms>]   Acquire a lock');
    console.log('  held                                  List locks held on this connection');
    console.log('  release <n>                           Release a held lock by its list number');
    console.log('  heartbeat                             Send a heartbeat for all held locks');
    console.log('  logout                                Revoke the current session');
    console.log('  quit                                  Exit');
}

async function handle(line) {
    const parts = line.trim().split(/\s+/).filter((p) => p.length > 0);
    if (parts.length === 0) return;
    const command = parts[0].toLowerCase();

    switch (command) {
        case 'help':
            printHelp();
            break;
        case 'login': {
            if (parts[1] === 'key' && parts.length >= 3) {
                const token = await admin.authenticateWithKey(parts[2]);
                tenantId = token.tenantId;
                console.log(`Authenticated. Tenant: ${token.tenantId}`);
            } else if (parts[1] === 'user' && parts.length >= 5) {
                const token = await admin.authenticateWithPassword(parts[2], parts[3], parts[4]);
                tenantId = token.tenantId;
                console.log(`Authenticated. Tenant: ${token.tenantId}`);
            } else {
                console.log('Usage: login key <accessKey>  |  login user <tenantId> <email> <password>');
            }
            break;
        }
        case 'whoami': {
            const d = await admin.getTokenDetails();
            console.log(`Principal : ${d.principalName}`);
            console.log(`Type      : ${d.principalType}`);
            console.log(`Tenant    : ${d.tenantId}`);
            console.log(`IsAdmin   : ${d.isAdmin}`);
            console.log(`TenantAdmin: ${d.isTenantAdmin}`);
            break;
        }
        case 'serverinfo': {
            const info = await admin.getServerInfo();
            console.log(`Product : ${info.product} ${info.version}`);
            console.log(`Node    : ${info.node}`);
            console.log(`Database: ${info.database}`);
            console.log(`WS conns: ${info.webSocketConnections}`);
            break;
        }
        case 'tenants': {
            const tenants = await admin.listTenants();
            console.log(`${tenants.length} tenant(s):`);
            for (const t of tenants) console.log(`  ${t.id}  ${t.name}  (active=${t.active}, protected=${t.isProtected})`);
            break;
        }
        case 'locks': {
            if (!tenantId) { console.log('Not authenticated. Use "login" first.'); break; }
            const holders = await admin.listLocks(tenantId);
            console.log(`${holders.length} active holder(s):`);
            for (const h of holders) console.log(`  ${h.lockKey}  ${h.mode}  fencing=${h.fencingToken}  holder=${h.id}`);
            break;
        }
        case 'connect': {
            if (parts.length < 2) { console.log('Usage: connect <accessKey>'); break; }
            if (locks) await locks.close();
            locks = new ClutchLockClient(endpoint, parts[1]);
            locks.on('heartbeat', (renewed) => console.log(`\n[push] heartbeat renewed ${renewed.length} lease(s)`));
            locks.on('error', (message) => console.log(`\n[push] error: ${message}`));
            locks.on('close', (reason) => console.log(`\n[push] connection closed: ${reason}`));
            const welcome = await locks.connect();
            console.log(`Connected. Session ${welcome.sessionId}, heartbeat every ${welcome.heartbeatIntervalMs} ms.`);
            break;
        }
        case 'acquire': {
            if (!locks) { console.log('Not connected. Use "connect" first.'); break; }
            if (parts.length < 3) { console.log('Usage: acquire <key> <Read|Write|Delete> [wait <ms>]'); break; }
            const options = {};
            if (parts[3] === 'wait' && parts[4]) { options.behavior = 'Wait'; options.timeoutMs = parseInt(parts[4], 10); }
            const acquired = await locks.acquire(parts[1], parts[2], options);
            const n = ++holderCounter;
            held.set(n, acquired);
            console.log(`Acquired [${n}] key=${acquired.key} mode=${acquired.mode} fencing=${acquired.fencingToken} lease=${acquired.leaseExpiresUtc}`);
            break;
        }
        case 'held': {
            if (held.size === 0) { console.log('No locks held on this connection.'); break; }
            for (const [n, h] of held) console.log(`  [${n}] key=${h.key} mode=${h.mode} fencing=${h.fencingToken} holder=${h.holderId}`);
            break;
        }
        case 'release': {
            if (!locks) { console.log('Not connected. Use "connect" first.'); break; }
            const n = parseInt(parts[1], 10);
            const h = held.get(n);
            if (!h) { console.log('Usage: release <n>  (see "held" for numbers)'); break; }
            const released = await locks.release(h.holderId);
            held.delete(n);
            console.log(`Released [${n}]: ${released}`);
            break;
        }
        case 'heartbeat': {
            if (!locks) { console.log('Not connected. Use "connect" first.'); break; }
            await locks.heartbeat();
            console.log('Heartbeat sent for all held locks.');
            break;
        }
        case 'logout': {
            await admin.logout();
            tenantId = null;
            console.log('Logged out.');
            break;
        }
        case 'quit':
        case 'exit':
            rl.close();
            return;
        default:
            console.log(`Unknown command '${command}'. Type 'help'.`);
    }
}

console.log('');
console.log('Clutch interactive console. Type "help" for commands, "quit" to exit.');
console.log(`Endpoint: ${endpoint}`);
console.log('');
rl.prompt();

rl.on('line', async (line) => {
    try {
        await handle(line);
    } catch (err) {
        if (err instanceof ClutchLockDeniedError) console.log(`Lock denied (${err.result}): ${err.message}`);
        else console.log(`Error: ${err.message}`);
    }
    rl.prompt();
});

rl.on('close', async () => {
    if (locks) await locks.close();
    console.log('Goodbye.');
    process.exit(0);
});
